[CORL-382] Allow moderators to ban a commenter from the comment stream

nick-funk opened the pull request:
## What does this PR do?

Adds a ban user drop down option in the comment stream's moderation drop down.

## How do I test this PR?

- Login as a moderate capable user
- Open the comment stream
- Find a comment with a user you want to ban
- Click the moderation drop down via the top right caret on the comment
- Click the new Ban User option
- Confirm you want to ban them
- User will be banned
- Click on caret again
- Drop down should have replaced the Ban User option with Banned

nick-funk committed:
Add a ban user action to the stream moderation drop down

CORL-382

nick-funk committed:
Show banned stated for user in stream moderation drop down

If the user is actively banned, the ban user option will be
disabled and show a status of "banned".

CORL-382

nick-funk committed:
Create utility for generating random stories, comments, and users

CORL-382

nick-funk committed:
Add ban and suspension values to baseUser in fixtures

CORL-382

nick-funk committed:
Updated banned.spec.tsx to use new test utilities for generating fixture data

CORL-382

nick-funk committed:
Prevent users from being able to ban themselves in the moderation dropdown

CORL-382

nick-funk committed:
Kill optimistic response errors for comment mutations

Set the author.status.current to an empty array so it stops
complaining about it being unused.

CORL-382

nick-funk committed:
Rename util in tests to helpers/fixture.ts

CORL-382

nick-funk committed:
Remove unused import from CreateCommentReplyMutation.ts

CORL-382

alinhle commented on src/core/client/stream/tabs/Comments/Comment/ReplyCommentForm/CreateCommentReplyMutation.ts:
```diff
@@ -185,7 +185,7 @@ function commit(
               ignoreable: false,
               status: {
-                current: [GQLUSER_STATUS.ACTIVE],
+                current: [],
```
Not sure if you are running into this: But be aware that Relay produces some false warnings that should be ignored. See https://github.com/facebook/relay/pull/2760

alinhle replied:
It will show a _false_ warning when putting scalar values inside an array in the optimistic response. Sth like 

```
Warning: validateMutation: `optimisticResponse` for mutation `ChangeTagsMutation`, contains an unused field ROOT.changeTags.tags.0
```

nick-funk replied:
I was indeed trying to silence these errors. I will put these back, as I don't want to be screwing with the optimistic response more than I have to. Even though they do spam the log with warnings.

nick-funk committed:
Put back the optimistic ban responses into comment mutations

The warnings spewing out during tests are false, for further
detail please see: https://github.com/facebook/relay/pull/2760

CORL-382

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-382

nick-funk committed:
Denormalize generated stories and comments

CORL-382

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/ModerationDropdown/ModerationActionsContainer.tsx:
```diff
@@ -0,0 +1,224 @@
+import RejectCommentMutation from "./RejectCommentMutation";
+import UnfeatureCommentMutation from "./UnfeatureCommentMutation";
+
+import cn from "classnames";
```
Import ordering.

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/ModerationDropdown/ModerationActionsContainer.tsx:
```diff
@@ -0,0 +1,224 @@
+      commentRevisionID: comment.revision.id,
+    });
+    onDismiss();
+  }, [feature, comment]);
```
Depends on `onDismiss`.

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/ModerationDropdown/ModerationActionsContainer.tsx:
```diff
@@ -0,0 +1,224 @@
+      <DropdownDivider />
+      <Localized id="comments-moderationDropdown-goToModerate">
+        <DropdownButton
+          href={`/admin/moderate/comment/${comment.id}`}
```
Note:

Investigate SSO implications.

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/ReplyCommentForm/CreateCommentReplyMutation.ts:
```diff
@@ -179,6 +184,12 @@ function commit(
+                ban: {
+                  active: false,
+                },
+              },
```
Replace with `lookup` based on the viewer.

nick-funk replied:
This should be fixed up now.

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/UserBanPopover/UserBanPopoverContainer.tsx:
```diff
@@ -0,0 +1,71 @@
+  const onBan = useCallback(() => {
+    banUser({ userID: user.id });
+    onDismiss();
+  }, [user, banUser]);
```
Missing `onDismiss` dep.

wyattjoh commented on src/core/client/stream/tabs/Comments/Stream/PostCommentForm/CreateCommentMutation.ts:
```diff
@@ -149,6 +149,12 @@ function commit(
+                ban: {
+                  active: false,
+                },
+              },
```
Replace with `lookup` based on viewer.

nick-funk committed:
Clean up import ordering in ModerationActionsContainer.tsx

CORL-382

nick-funk committed:
Inject appropriate scoped items into callbacks for moderation dropdown

CORL-382

nick-funk committed:
Set optimistic response author status from known viewer status

CORL-382

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-382

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-382
